### ISSUE TITLE
scaling up ceph mon count doesn't always work

### ISSUE THREAD

**Issue by jtlayton** [rook/rook#2645]
I've been using the files under cluster/examples/kubernetes/ceph to bring up rook clusters, and I've found that sometimes when I try to scale out the mon count after starting the cluster, it doesn't work.

Start by modifying the files under cluster/examples/kubernetes/ceph to start only a single monitor and a manager daemon. Set useAllNodes and useAllDevices to false so that no osd is created. Start up the cluster and the toolbox container.

Exec a shell in the toolbox and run : `ceph orchestrator set backend rook`

...to work around the race condition here: https://github.com/rook/rook/issues/2617

On my single-node k8s cluster (kube.poochiereds.net), I've dedicated a block device for the OSD (/dev/vdb). Use the orchestrator to create the block device. In the toolbox container shell:

`ceph orchestrator osd create kube.poochiereds.net:vdb`

After that starts, edit the cephcluster object:

`kubectl edit -n rook-ceph cephcluster rook-ceph`

...change the spec.mon.count field from 1 to 3 and save the result. The  mon healthcheck should eventually run and spin up the other two mons, but that never occurs on my setup.

If I skip creating the OSD, then it seems to work correctly.

**Comment by jtlayton** [comment]
@travisn asked for the operator log:

[operator-log.txt](https://github.com/rook/rook/files/2857711/operator-log.txt)

**Comment by jtlayton** [comment]
Debug level logging. Looks like the health check ran a couple of times after I edited the mon count:

[operator-debug.txt](https://github.com/rook/rook/files/2857795/operator-debug.txt)

**Comment by travisn** [comment]
The changing value from 1 to 3 seems to be lost by the health monitoring goroutine. The critical part of the log:
```
2019-02-12 21:42:45.990169 I | op-cluster: number of mons have changed from 1 to 3. The health check will update the mons...
2019-02-12 21:42:45.990685 I | op-cluster: update event for cluster rook-ceph is not supported
2019-02-12 21:42:53.458680 D | op-mon: checking health of mons
2019-02-12 21:42:53.458793 D | op-mon: Checking health for mons (desired=1). 
```

**Comment by jtlayton** [comment]
Doing a bit of printf debugging here. This is what I see when I try to scale up the mon count after adding an OSD:
```
$ kubectl logs -n rook-ceph-system rook-ceph-operator-5c64657b79-2gzv8 | egrep 'have changed|Checking health'
2019-03-01 14:33:00.169934 D | op-mon: Checking health for mons (cluster=0xc000b73e40 desired=1). &{FSID:58812736-0739-441a-93bc-ec5e03c07aec MonitorSecret:AQAuQnlcTA5VBRAA6u0qVEXR3wOzmJYMqIxDnw== AdminSecret:AQAuQnlc8ppcCBAAL91Iau6fTHAp+v8L+J0Jyw== Name:rook-ceph Monitors:map[a:0xc000cf72a0] CephVersionName:nautilus}
2019-03-01 14:33:45.640068 D | op-mon: Checking health for mons (cluster=0xc000b73e40 desired=1). &{FSID:58812736-0739-441a-93bc-ec5e03c07aec MonitorSecret:AQAuQnlcTA5VBRAA6u0qVEXR3wOzmJYMqIxDnw== AdminSecret:AQAuQnlc8ppcCBAAL91Iau6fTHAp+v8L+J0Jyw== Name:rook-ceph Monitors:map[a:0xc000cf72a0] CephVersionName:nautilus}
2019-03-01 14:34:31.113131 D | op-mon: Checking health for mons (cluster=0xc000b73e40 desired=1). &{FSID:58812736-0739-441a-93bc-ec5e03c07aec MonitorSecret:AQAuQnlcTA5VBRAA6u0qVEXR3wOzmJYMqIxDnw== AdminSecret:AQAuQnlc8ppcCBAAL91Iau6fTHAp+v8L+J0Jyw== Name:rook-ceph Monitors:map[a:0xc000cf72a0] CephVersionName:nautilus}
2019-03-01 14:35:04.191463 I | op-cluster: number of mons in 0xc000af2580 have changed from 1 to 3. The health check will update the mons...
2019-03-01 14:35:16.593384 D | op-mon: Checking health for mons (cluster=0xc000b73e40 desired=1). &{FSID:58812736-0739-441a-93bc-ec5e03c07aec MonitorSecret:AQAuQnlcTA5VBRAA6u0qVEXR3wOzmJYMqIxDnw== AdminSecret:AQAuQnlc8ppcCBAAL91Iau6fTHAp+v8L+J0Jyw== Name:rook-ceph Monitors:map[a:0xc000cf72a0] CephVersionName:nautilus}
```
The pointer value in the op-cluster line is different from the ones in the op-mon line. If I do the scaling before adding an OSD, it's the same value.

Looking over the code, it looks like the healthCheck is spawned in onAdd(), but the createInstance() call is also done in onUpdate(). The cluster object then ends up getting a new "mon" cluster, but the health check is never updated to follow the new one.

Maybe spawning the health check should be done in createInstance?

**Comment by travisn** [comment]
When I was looking at #2740 last night, I found some references to c.Count that may be causing these issues. Instead of the health check referencing the c.Count, I think we should set variables on an independent health check object so that the constantly changing crd wouldn't impact the values the health check is working on.

**Comment by jtlayton** [comment]
That sounds sensible.

**Comment by travisn** [comment]
This should already be fixed with #2835
